Stream Enrich: add the possibility to override the DynamoDB endpoint (closes #3942)

szareiangm opened the pull request:
- We have added the optional endpoint to Dynamodb to make it work with Localstack.
- Also, Cloudwatch metrics was not behaving as expected with Localstack so we needed to use the NullFactory to disable it. 

<!--
Thank you for contributing to Snowplow!

You'll find a small checklist below which should help speed up the review processs:

- [ ] Have you signed the [contributor license agreement](https://github.com/snowplow/snowplow/wiki/CLA)?
- [ ] Have you read the [contributing guide](https://github.com/snowplow/snowplow/blob/master/CONTRIBUTING.md)?
- [ ] Have you added the appropriate unit tests?
- [ ] Have you run the tests through `sbt test`?
- [ ] Is your pull request against the `master` branch?
-->

szareiangm commented:
Not sure what has failed in the build. can someone please restart it?

chuwy commented:
@szareiangm done.

benjben commented on 3-enrich/stream-enrich/core/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/Amodel.scala:
```diff
@@ -64,6 +66,17 @@ object model {
+      case _                 => s"https://dynamodb.$region.amazonaws.com"
+    })
+
+    val iscloudwatchDisabled = disableCloudWatch.getOrElse(region match {
```
You don't need this.

szareiangm replied:
removed.

benjben commented on 3-enrich/stream-enrich/kinesis/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/sources/KinesisSource.scala:
```diff
@@ -140,11 +142,18 @@ class KinesisSource private (
-      kinesisClientLibConfiguration
-    )
-
+    val worker = if (kinesisConfig.disableCloudWatch.isDefined && kinesisConfig.iscloudwatchDisabled == true){
```
You can directly use `kinesisConfig.disableCloudWatch` :
```
val worker = kinesisConfig.disableCloudWatch match {
  case Some(true) =>
    new Worker(
      rawEventProcessorFactory,
      kinesisClientLibConfiguration,
      new NullMetricsFactory()
    )
  case _ =>
    new Worker(
      rawEventProcessorFactory,
      kinesisClientLibConfiguration
    )
}
```

szareiangm replied:
done

benjben commented on 3-enrich/stream-enrich/examples/config.hocon.sample:
```diff
@@ -69,6 +69,13 @@ enrich {
       # customEndpoint = {{kinesisEndpoint}}
       # customEndpoint = ${?ENRICH_STREAMS_SOURCE_SINK_CUSTOM_ENDPOINT}
 
+      ## Optional endpoint url configuration to override aws dyanomdb endpoints,
```
Can you please precise that this is for Kinesis checkpoints.

szareiangm replied:
done

benjben commented on 3-enrich/stream-enrich/examples/config.hocon.sample:
```diff
@@ -69,6 +69,13 @@ enrich {
+      # dynamodbCustomEndpoint = "http://localhost:4569"
+
+      # Optional override to disable cloudwatch; set to true to disable
+      # disableCloudWatch = false
```
=>
```
# Optional override to disable cloudwatch
# disableCloudWatch = true
```

szareiangm replied:
done

szareiangm commented:
I have addressed all of the comments. This works is part of the project that our Summer co-op student, Saud Naumani, has done. I just submitted it and addressed the comments because he went back to the university.

szareiangm commented:
also, you might want to run `scalafmt` on stream-enrich for the next release preparation commit. I could run it but the files are too many for the purpose of this ticket.

benjben commented:
@szareiangm could you please update this PR so that it can be merged into `release/r119` ?

szareiangm commented:
Hi, sure. I will give it a shot until Monday.

benjben commented:
Great thanks!

benjben commented:
Hi @szareiangm ,
As we are going to release R119 soon and we'd like this feature to be in it, I reapplied the changes manually on another branch and used your name and email as author : https://github.com/snowplow/snowplow/pull/4327
Thanks for the contribution!
